// react-native-qr-svg/lib/module/QrCodeSvg.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import React, { useMemo, useRef } from 'react';
import { createMatrix } from './createMatrix';
import Svg, { Defs, G, LinearGradient, Path, Rect, Stop } from 'react-native-svg';
import { StyleSheet, View } from 'react-native';
import renderFigure from './renderFigure';
import { defaultRenderer } from './renderers';
import { nanoid } from 'nanoid/non-secure';
import { round } from './round';

/**
 * Properties for configuring the SVG QR code generation component.
 */

export default function QrCodeSvg(_ref) {
  let {
    value,
    frameSize,
    contentCells = 6,
    errorCorrectionLevel = 'M',
    backgroundColor = '#ffffff',
    dotColor = '#000000',
    style,
    contentBackgroundRectProps,
    content,
    contentStyle,
    figureCircleProps,
    figurePathProps,
    renderer = defaultRenderer,
    gradientColors,
    gradientProps
  } = _ref;
  const originalMatrix = useMemo(() => createMatrix(value, errorCorrectionLevel), [errorCorrectionLevel, value]);
  const cellSize = round(frameSize / originalMatrix.length); // Ex. 3.141592653589793 -> 3.14
  const matrixRowLength = originalMatrix[0]?.length ?? 0;
  const roundedContentCells = (matrixRowLength - contentCells) % 2 === 0 ? contentCells : contentCells + 1;
  const contentSize = round(cellSize * roundedContentCells);
  const contentStartIndex = (matrixRowLength - roundedContentCells) / 2;
  const contentEndIndex = contentStartIndex + roundedContentCells - 1;
  const contentXY = contentStartIndex * cellSize;
  const matrix = useMemo(() => content !== undefined ? originalMatrix.map((row, i) => row.map((el, j) => i >= contentStartIndex && i <= contentEndIndex && j >= contentStartIndex && j <= contentEndIndex ? 0 : el)) : originalMatrix, [content, contentEndIndex, contentStartIndex, originalMatrix]);
  const matrixFocusSquareDeepIndex = matrix[0]?.findIndex(_ => _ === 0);
  if (matrixFocusSquareDeepIndex === undefined) {
    throw new Error("Focus square wasn't detected");
  }
  const matrixFocusSquareDeep = matrixFocusSquareDeepIndex;
  const paths = useMemo(() => matrix.flatMap((row, i) => row.flatMap((_, j) => {
    if (!row?.[j]) {
      return [];
    }
    const neighbors = {
      top: Boolean(matrix[i - 1]?.[j]),
      bottom: Boolean(matrix[i + 1]?.[j]),
      left: Boolean(row[j - 1]),
      right: Boolean(row[j + 1])
    };
    const x = j * cellSize;
    const y = i * cellSize;
    return [renderFigure(x, y, neighbors, cellSize, renderer, matrixFocusSquareDeep, i, j, matrix.length)];
  })), [matrix, cellSize, renderer, matrixFocusSquareDeep]);
  const dPath = paths.filter(_ => _.type === 'path').map(_ => _.d).join(' ');
  const dCircle = paths.filter(_ => _.type === 'circle').map(_ => _.d).join(' ');
  const qrProps = {
    frameSize,
    dPath,
    dCircle,
    figurePathProps,
    figureCircleProps,
    backgroundColor,
    contentXY,
    contentBackgroundRectProps,
    contentSize,
    dotColor
  };
  return /*#__PURE__*/React.createElement(View, {
    testID: "root",
    style: [{
      backgroundColor
    }, style]
  }, /*#__PURE__*/React.createElement(View, null, Array.isArray(gradientColors) ? /*#__PURE__*/React.createElement(GradientQr, _extends({}, qrProps, {
    gradientColors: gradientColors,
    gradientProps: gradientProps
  })) : /*#__PURE__*/React.createElement(DefaultQr, qrProps), content && /*#__PURE__*/React.createElement(View, {
    testID: "content",
    style: [{
      width: contentSize,
      height: contentSize,
      top: contentXY,
      left: contentXY
    }, styles.content, contentStyle]
  }, content)));
}
const DefaultQr = _ref2 => {
  let {
    frameSize,
    dPath,
    dCircle,
    figurePathProps,
    figureCircleProps,
    contentXY,
    content,
    contentSize,
    backgroundColor,
    contentBackgroundRectProps,
    dotColor
  } = _ref2;
  return /*#__PURE__*/React.createElement(Svg, {
    testID: "svg",
    width: frameSize,
    height: frameSize
  }, /*#__PURE__*/React.createElement(G, null, /*#__PURE__*/React.createElement(Path, _extends({
    testID: "dot",
    d: dPath,
    fill: dotColor
  }, figurePathProps)), /*#__PURE__*/React.createElement(Path, _extends({
    testID: "dot",
    d: dCircle,
    fill: dotColor
  }, figureCircleProps))), content && /*#__PURE__*/React.createElement(Rect, _extends({
    fill: backgroundColor,
    x: contentXY,
    y: contentXY
  }, contentBackgroundRectProps, {
    width: contentSize,
    height: contentSize
  })));
};
const GradientQr = _ref3 => {
  let {
    frameSize,
    dPath,
    dCircle,
    figurePathProps,
    figureCircleProps,
    contentXY,
    content,
    contentSize,
    backgroundColor,
    contentBackgroundRectProps,
    gradientColors,
    gradientProps
  } = _ref3;
  const id = useRef(nanoid(10)).current;
  return /*#__PURE__*/React.createElement(Svg, {
    testID: "svg",
    width: frameSize,
    height: frameSize
  }, /*#__PURE__*/React.createElement(Defs, null, /*#__PURE__*/React.createElement(LinearGradient, _extends({
    id: id,
    x1: "0",
    y1: "0",
    x2: "1",
    y2: "1"
  }, gradientProps), /*#__PURE__*/React.createElement(Stop, {
    offset: "0",
    stopColor: gradientColors[0]
  }), /*#__PURE__*/React.createElement(Stop, {
    offset: "1",
    stopColor: gradientColors[1]
  }))), /*#__PURE__*/React.createElement(G, null, /*#__PURE__*/React.createElement(Path, _extends({
    d: dPath,
    fill: `url(#${id})`
  }, figurePathProps)), /*#__PURE__*/React.createElement(Path, _extends({
    d: dCircle,
    fill: `url(#${id})`
  }, figureCircleProps))), content && /*#__PURE__*/React.createElement(Rect, _extends({
    fill: backgroundColor,
    x: contentXY,
    y: contentXY
  }, contentBackgroundRectProps, {
    width: contentSize,
    height: contentSize
  })));
};
const styles = StyleSheet.create({
  content: {
    position: 'absolute'
  }
});
