// react-native-qr-svg/lib/module/renderers.js
import { Kind } from './types';
import { round } from './round';
const pair = d => `${d.x} ${d.y}`;
const line = d => `L${pair(d)}`;
export const plainRenderer = {
  render: {
    [Kind.Circle]: _ref => {
      let {
        corners: c
      } = _ref;
      return renderSquare(c);
    },
    [Kind.Element]: _ref2 => {
      let {
        corners: c
      } = _ref2;
      return renderSquare(c);
    }
  }
};
export const defaultRenderer = {
  render: {
    [Kind.Circle]: _ref3 => {
      let {
        corners,
        cellSize
      } = _ref3;
      return renderCircle(corners.center, cellSize);
    },
    [Kind.Element]: _ref4 => {
      let {
        neighbors,
        corners
      } = _ref4;
      const {
        q2,
        q3,
        q4,
        q1,
        d1,
        d2,
        d4,
        d3
      } = corners;
      // q4  0  d1  0  q1
      // 0   0  0   0  0
      // d4  0  0   0  d2
      // 0   0  0   0  0
      // q3  0  d3  0  q2
      const d1d2 = neighbors.top || neighbors.right ? `L${q1.x} ${q1.y} L${d2.x} ${d2.y}` : `Q${q1.x} ${q1.y} ${d2.x} ${d2.y}`;
      const d2d3 = neighbors.right || neighbors.bottom ? `L${q2.x} ${q2.y} L${d3.x} ${d3.y}` : `Q${q2.x} ${q2.y} ${d3.x} ${d3.y}`;
      const d3d4 = neighbors.bottom || neighbors.left ? `L${q3.x} ${q3.y} L${d4.x} ${d4.y}` : `Q${q3.x} ${q3.y} ${d4.x} ${d4.y}`;
      const d4d1 = neighbors.left || neighbors.top ? `L${q4.x} ${q4.y} L${d1.x} ${d1.y}` : `Q${q4.x} ${q4.y} ${d1.x} ${d1.y}`;
      return `M${d1.x} ${d1.y} ${d1d2} ${d2d3} ${d3d4} ${d4d1}`;
    }
  }
};
export const triangleRenderer = {
  render: {
    [Kind.Circle]: _ref5 => {
      let {
        corners: c
      } = _ref5;
      return `M${pair(c.d1)} ${line(c.d2)} ${line(c.d3)} ${line(c.d4)}`;
    },
    [Kind.Element]: _ref6 => {
      let {
        corners: c
      } = _ref6;
      return `M${pair(c.d1)} ${line(c.d2)} ${line(c.d3)} ${line(c.d4)}`;
    }
  }
};
export const circleRenderer = {
  render: {
    [Kind.Circle]: _ref7 => {
      let {
        corners,
        cellSize
      } = _ref7;
      return renderCircle(corners.center, cellSize);
    },
    [Kind.Element]: _ref8 => {
      let {
        corners,
        cellSize
      } = _ref8;
      return renderCircle(corners.center, cellSize);
    }
  }
};
export const renderCircle = (center, cellSize) => {
  const half = round(cellSize / 2);
  return `M${center.x + half} ${center.y} A${half} ${half} 0 1 0 ${center.x - half} ${center.y} A${half} ${half} 0 1 0 ${center.x + half} ${center.y}`;
};
export const renderSquare = c => `M${pair(c.q1)} ${line(c.q2)} ${line(c.q3)} ${line(c.q4)}`;
