// react-native-qr-svg/lib/commonjs/QrCodeSvg.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = QrCodeSvg;
var _react = _interopRequireWildcard(require("react"));
var _createMatrix = require("./createMatrix");
var _reactNativeSvg = _interopRequireWildcard(require("react-native-svg"));
var _reactNative = require("react-native");
var _renderFigure = _interopRequireDefault(require("./renderFigure"));
var _renderers = require("./renderers");
var _nonSecure = require("nanoid/non-secure");
var _round = require("./round");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function QrCodeSvg(_ref) {
  let {
    value,
    frameSize,
    contentCells = 6,
    errorCorrectionLevel = 'M',
    backgroundColor = '#ffffff',
    dotColor = '#000000',
    style,
    contentBackgroundRectProps,
    content,
    contentStyle,
    figureCircleProps,
    figurePathProps,
    renderer = _renderers.defaultRenderer,
    gradientColors,
    gradientProps
  } = _ref;
  const originalMatrix = (0, _react.useMemo)(() => (0, _createMatrix.createMatrix)(value, errorCorrectionLevel), [errorCorrectionLevel, value]);
  const cellSize = (0, _round.round)(frameSize / originalMatrix.length); // Ex. 3.141592653589793 -> 3.14
  const matrixRowLength = originalMatrix[0]?.length ?? 0;
  const roundedContentCells = (matrixRowLength - contentCells) % 2 === 0 ? contentCells : contentCells + 1;
  const contentSize = (0, _round.round)(cellSize * roundedContentCells);
  const contentStartIndex = (matrixRowLength - roundedContentCells) / 2;
  const contentEndIndex = contentStartIndex + roundedContentCells - 1;
  const contentXY = contentStartIndex * cellSize;
  const matrix = (0, _react.useMemo)(() => content !== undefined ? originalMatrix.map((row, i) => row.map((el, j) => i >= contentStartIndex && i <= contentEndIndex && j >= contentStartIndex && j <= contentEndIndex ? 0 : el)) : originalMatrix, [content, contentEndIndex, contentStartIndex, originalMatrix]);
  const matrixFocusSquareDeepIndex = matrix[0]?.findIndex(_ => _ === 0);
  if (matrixFocusSquareDeepIndex === undefined) {
    throw new Error("Focus square wasn't detected");
  }
  const matrixFocusSquareDeep = matrixFocusSquareDeepIndex;
  const paths = (0, _react.useMemo)(() => matrix.flatMap((row, i) => row.flatMap((_, j) => {
    if (!row?.[j]) {
      return [];
    }
    const neighbors = {
      top: Boolean(matrix[i - 1]?.[j]),
      bottom: Boolean(matrix[i + 1]?.[j]),
      left: Boolean(row[j - 1]),
      right: Boolean(row[j + 1])
    };
    const x = j * cellSize;
    const y = i * cellSize;
    return [(0, _renderFigure.default)(x, y, neighbors, cellSize, renderer, matrixFocusSquareDeep, i, j, matrix.length)];
  })), [matrix, cellSize, renderer, matrixFocusSquareDeep]);
  const dPath = paths.filter(_ => _.type === 'path').map(_ => _.d).join(' ');
  const dCircle = paths.filter(_ => _.type === 'circle').map(_ => _.d).join(' ');
  const qrProps = {
    frameSize,
    dPath,
    dCircle,
    figurePathProps,
    figureCircleProps,
    backgroundColor,
    contentXY,
    contentBackgroundRectProps,
    contentSize,
    dotColor
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    testID: "root",
    style: [{
      backgroundColor
    }, style]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, null, Array.isArray(gradientColors) ? /*#__PURE__*/_react.default.createElement(GradientQr, _extends({}, qrProps, {
    gradientColors: gradientColors,
    gradientProps: gradientProps
  })) : /*#__PURE__*/_react.default.createElement(DefaultQr, qrProps), content && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    testID: "content",
    style: [{
      width: contentSize,
      height: contentSize,
      top: contentXY,
      left: contentXY
    }, styles.content, contentStyle]
  }, content)));
}
const DefaultQr = _ref2 => {
  let {
    frameSize,
    dPath,
    dCircle,
    figurePathProps,
    figureCircleProps,
    contentXY,
    content,
    contentSize,
    backgroundColor,
    contentBackgroundRectProps,
    dotColor
  } = _ref2;
  return /*#__PURE__*/_react.default.createElement(_reactNativeSvg.default, {
    testID: "svg",
    width: frameSize,
    height: frameSize
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.G, null, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Path, _extends({
    testID: "dot",
    d: dPath,
    fill: dotColor
  }, figurePathProps)), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Path, _extends({
    testID: "dot",
    d: dCircle,
    fill: dotColor
  }, figureCircleProps))), content && /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Rect, _extends({
    fill: backgroundColor,
    x: contentXY,
    y: contentXY
  }, contentBackgroundRectProps, {
    width: contentSize,
    height: contentSize
  })));
};
const GradientQr = _ref3 => {
  let {
    frameSize,
    dPath,
    dCircle,
    figurePathProps,
    figureCircleProps,
    contentXY,
    content,
    contentSize,
    backgroundColor,
    contentBackgroundRectProps,
    gradientColors,
    gradientProps
  } = _ref3;
  const id = (0, _react.useRef)((0, _nonSecure.nanoid)(10)).current;
  return /*#__PURE__*/_react.default.createElement(_reactNativeSvg.default, {
    testID: "svg",
    width: frameSize,
    height: frameSize
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Defs, null, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.LinearGradient, _extends({
    id: id,
    x1: "0",
    y1: "0",
    x2: "1",
    y2: "1"
  }, gradientProps), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Stop, {
    offset: "0",
    stopColor: gradientColors[0]
  }), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Stop, {
    offset: "1",
    stopColor: gradientColors[1]
  }))), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.G, null, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Path, _extends({
    d: dPath,
    fill: `url(#${id})`
  }, figurePathProps)), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Path, _extends({
    d: dCircle,
    fill: `url(#${id})`
  }, figureCircleProps))), content && /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Rect, _extends({
    fill: backgroundColor,
    x: contentXY,
    y: contentXY
  }, contentBackgroundRectProps, {
    width: contentSize,
    height: contentSize
  })));
};
const styles = _reactNative.StyleSheet.create({
  content: {
    position: 'absolute'
  }
});
